In [29]:
from pathlib import Path

import pandas as pd

pd.options.display.max_columns = None

# The importance of sandboxes

I've worked with SQL regularly over the past decade, but over the first half of that decade (and despite extensive formal knowledge of normalization, data modeling, and other database-related topics) I essentially only used SQL to extract data and use Python or R to do analysis. I only had access to production databases and I didn't know enough about how database management system work to risk experiments, and as a result, I learned at a glacial pace. But in 2018, I stumbled into a project interesting enough to motivate me to install PostGIS on a personal machine and freed from the fear of accidentally taking down a production system (and the power of a superuser), I was able to lift up the hood and see how the machine worked, and this enabled me to learn and build with PostgreSQL + PostGIS (a geospatial PostgreSQL extension) far faster.

In this post, I show how to use Docker to set up a PostGIS database and experiment with it.

## What is Docker

Docker and Docker Compose[1^] enable you to import and run complicated applications in just a few lines of code.

The internals of docker are really interesting, but for this post, all you need to know these things:

1. A docker **image** is like a blueprint of an application, and it's defined in **Dockerfiles**.
2. A docker **container** is a runnable instance of the application, built from the instructions in the blueprint.
3. You can configure your application in a `docker-compose.yml` file.

### Our Dockerfile
The first one may sound complicated, as the contents of our Dockerfile show, this can only take one line of code (from us). That's possible because Dockerfiles can build an image based on another image, and the developers of many open source projects (e.g. Ubuntu, PostgreSQL/PostGIS, MySQL, Go, nginx, etc) public official images on [Docker Hub](https://registry.hub.docker.com/search?q=). Our Dockerfile pulls the [postgis/postgis image](https://registry.hub.docker.com/r/postgis/postgis) (translated, it pulls the `postgis` base image from the `postgis` organization), and then adds nothing else. So in one line, we indicate that we want the `postgis/postgis` image that has the tag "15-3.3" (which they've intuitively used to indicate the image provides a PostGIS database that has version 15-3.3).

In [2]:
#| code-fold: true
#| code-summary: The full contents of our Dockerfile

!cat db_context/Dockerfile

FROM postgis/postgis:15-3.3

### Our docker-compose.yml file

And this is our `docker-compose.yml` file. The `docker-compose.yml` file defines the services, and any networks, volumes, configs, and/or secrets/environment variables your system needs to work. The file below defines one service (named `postgis`) and one volume (named `sandbox_postgis_data`).

In [3]:
#| code-fold: true
#| code-summary: The contents of our docker-compose.yml

!cat docker-compose.yml

version: '3.9'

services:
  postgis:
    image: sandbox_postgis:15.3.3
    build:
      context: ./db_context
      dockerfile: Dockerfile
    ports:
      - 54321:5432
    environment:
      POSTGRES_DB: db_name
      POSTGRES_USER: db_username
      POSTGRES_PASSWORD: db_password
    volumes:
      - sandbox_postgis_data:/var/lib/postgresql/data

volumes:
  sandbox_postgis_data:


There are 3 top level elements in this file: `version`, `services`, and `volumes`.

* `version` indicates the version of the Docker Compose specification; I don't think I've ever set it to anything other than 3.9.
* `services` defines configurations for each container your system needs.
* `volumes` defines persistant data stores that can be shared by different services.

#### Services
The `postgis` service has five elements: `image`, `build`, `ports`, `environment`, and `volumes`.

The `image` and `build` elements define the docker image to use; `image` defines both the name ("sandbox_postgis") and tag ("15.5.3") for the docker image, and `build` defines the Dockerfile to build into an image as well as the context to build into the docker image.

The `port` element defines a connection from a port on the host machine (host post 54321) to a port into the container (container port 5432, the default for PostgreSQL). We'll use that later to connect to the database.

The `volumes` element (in the `postgis` service) defines a persistant storage volume that will hold the data in our database. Without this, our database would reset every time we restart the system [^2].

And the `environment` element enables you to set environment variables in the container. Here, we pass in environment variables `POSTGRES_DB`, `POSTGRES_USER`, and `POSTGRES_PASSWORD` which are used to name the database (in this case, it's named `db_name`) and create a superuser (in this case having username `db_username` and password `db_password`) when the database is first created in a new volume.

### Building the image(s) used in our system

The first time you build the image defined by your Dockerfile, docker will read your Dockerfile(s), download all layers of the base image(s) (defined in lines starting with `FROM `), process each subsequent instruction into a layer, cache layers, and then bind the layers into an image. This is a template for creating containers.

The second time you run this command (assuming no changes have been made to the `docker-compose.yml` file, Dockerfile(s), or any other files the Dockerfile references), all layers will just be pulled from cache, producing much smaller output (like what's shown below).

In [4]:
#| code-fold: true
#| code-summary: The output produced while building our images

!docker compose build

[+] Building 0.0s (0/1)                                                         
[+] Building 0.1s (2/4)                                                         
 => [postgis internal] load build definition from Dockerfile               0.0s
 => => transferring dockerfile: 64B                                        0.0s
 => [postgis internal] load .dockerignore                                  0.0s
 => => transferring context: 2B                                            0.0s
 => [postgis internal] load metadata for docker.io/postgis/postgis:15-3.3  0.1s
 => [postgis auth] postgis/postgis:pull token for registry-1.docker.io     0.0s
[+] Building 0.3s (3/4)                                                         
 => [postgis internal] load build definition from Dockerfile               0.0s
 => => transferring dockerfile: 64B                                        0.0s
 => [postgis internal] load .dockerignore                                  0.0s
 => => transferring context: 2B      

In [38]:
#| code-fold: true
#| code-summary: Shows that our system's one container is not running yet

!docker ps -f name=sandbox-postgis

CONTAINER ID   IMAGE     COMMAND   CREATED   STATUS    PORTS     NAMES


`docker ps` shows running containers, and the `-f name=...` option allows us to filter to running containers with a name containing the entered string. Currently, "sandbox-postgis" isn't part of the name of any running container, so let's spin one up.

In [39]:
#| code-fold: false
#| code-summary: The command to start up our system

!docker compose up -d

[+] Running 1/0
 ✔ Network 015_docker_postgres_sandbox_default      Created                0.1s 
 ⠋ Container 015_docker_postgres_sandbox-postgis-1  Creating               0.0s 
[+] Running 1/2
 ✔ Network 015_docker_postgres_sandbox_default      Created                0.1s 
 ⠙ Container 015_docker_postgres_sandbox-postgis-1  Creating               0.1s 
[+] Running 1/2
 ✔ Network 015_docker_postgres_sandbox_default      Created                0.1s 
 ⠹ Container 015_docker_postgres_sandbox-postgis-1  Creating               0.2s 
[+] Running 1/2
 ✔ Network 015_docker_postgres_sandbox_default      Created                0.1s 
 ⠸ Container 015_docker_postgres_sandbox-postgis-1  Creating               0.3s 
[+] Running 1/2
 ✔ Network 015_docker_postgres_sandbox_default      Created                0.1s 
 ⠼ Container 015_docker_postgres_sandbox-postgis-1  Creating               0.4s 
[+] Running 1/2
 ✔ Network 015_docker_postgres_sandbox_default      Created                0.1s 
 ⠴ Container 

In [40]:
#| code-fold: true
#| code-summary: Shows that our system's one container is running now

!docker ps -f name=sandbox-postgis

CONTAINER ID   IMAGE                    COMMAND                  CREATED         STATUS                  PORTS                                         NAMES
8513cc1ff07d   sandbox_postgis:15.3.3   "docker-entrypoint.s…"   2 seconds ago   Up Less than a second   0.0.0.0:54321->5432/tcp, :::54321->5432/tcp   015_docker_postgres_sandbox-postgis-1


## Sending commands to our database

Now we have our database up and running on our local system, but now we need to connect to it. There are countless ways to connect to databases, but I want to focus on the SQL, so I've written just enough to allow me to connect to our datasbase (via my `get_db_connection()` function) and execute a SQL query/statement (via my `execute_query()` function.

Note that `get_db_connection()` uses the `POSTGRES_DB`, `POSTGRES_USER`, and `POSTGRES_PASSWORD` environment variables as well as port number 54321, all of which we set in our `docker-compose.yml` file.

In [41]:
#| code-fold: false
#| code-summary: Our extremely thi

from contextlib import contextmanager
from typing import Union

import pandas as pd
import psycopg2


@contextmanager
def get_db_connection():
    conn = psycopg2.connect(
        dbname="db_name",
        user="db_username",
        password="db_password",
        host="localhost",
        port=54321
    )
    try:
        yield conn
    finally:
        conn.close()

def execute_query(query: str) -> Union[pd.DataFrame, None]:
    with get_db_connection() as conn:
        with conn.cursor() as cur:
            cur.execute(query)
            rows = cur.fetchall()
            if rows:
                col_names = [desc[0] for desc in cur.description]
                return pd.DataFrame(rows, columns=col_names)

In [20]:
schemas = execute_query(query="""
    SELECT nspname AS schema_name
    FROM pg_namespace;
""")
schemas

schema_name
0            pg_toast
1          pg_catalog
2              public
3  information_schema
4            topology
5               tiger
6          tiger_data

In [23]:
execute_query(query="""SELECT current_user;""")

current_user
0  db_username

In [24]:
pg_databases_df = execute_query(
    query="""
        SELECT *
        FROM pg_database;
    """
)
pg_databases_df

oid           datname  datdba  encoding datlocprovider  datistemplate  \
0      5          postgres      10         6              c          False   
1  16384           db_name      10         6              c          False   
2      1         template1      10         6              c           True   
3      4         template0      10         6              c           True   
4  16385  template_postgis      10         6              c           True   

   datallowconn  datconnlimit datfrozenxid datminmxid  dattablespace  \
0          True            -1          717          1           1663   
1          True            -1          717          1           1663   
2          True            -1          717          1           1663   
3         False            -1          717          1           1663   
4          True            -1          717          1           1663   

   datcollate    datctype daticulocale datcollversion  \
0  en_US.utf8  en_US.utf8         None           2.31   
1  en_US.utf8  en_US.utf8         None           2.31   
2  en_US.utf8  en_US.utf8         None           2.31   
3  en_US.utf8  en_US.utf8         None           None   
4  en_US.utf8  en_US.utf8         None           2.31   

                                         datacl  
0                                          None  
1                                          None  
2  {=c/db_username,db_username=CTc/db_username}  
3  {=c/db_username,db_username=CTc/db_username}  
4                                          None

In [25]:
pg_namespaces_df = execute_query(
    query="""
        SELECT *
        FROM pg_namespace;
    """
)
pg_namespaces_df

oid             nspname  nspowner  \
0     99            pg_toast        10   
1     11          pg_catalog        10   
2   2200              public      6171   
3  13186  information_schema        10   
4  19052            topology        10   
5  19226               tiger        10   
6  19482          tiger_data        10   

                                              nspacl  
0                                               None  
1        {db_username=UC/db_username,=U/db_username}  
2  {pg_database_owner=UC/pg_database_owner,=U/pg_...  
3        {db_username=UC/db_username,=U/db_username}  
4        {db_username=UC/db_username,=U/db_username}  
5                                               None  
6                                               None

In [26]:
pg_roles_df = execute_query(
    query="""
        SELECT *
        FROM pg_roles;
    """
)
pg_roles_df

rolname  rolsuper  rolinherit  rolcreaterole  \
0           pg_database_owner     False        True          False   
1            pg_read_all_data     False        True          False   
2           pg_write_all_data     False        True          False   
3                  pg_monitor     False        True          False   
4        pg_read_all_settings     False        True          False   
5           pg_read_all_stats     False        True          False   
6         pg_stat_scan_tables     False        True          False   
7        pg_read_server_files     False        True          False   
8       pg_write_server_files     False        True          False   
9   pg_execute_server_program     False        True          False   
10          pg_signal_backend     False        True          False   
11              pg_checkpoint     False        True          False   
12                db_username      True        True           True   

    rolcreatedb  rolcanlogin  rolreplication  rolconnlimit rolpassword  \
0         False        False           False            -1    ********   
1         False        False           False            -1    ********   
2         False        False           False            -1    ********   
3         False        False           False            -1    ********   
4         False        False           False            -1    ********   
5         False        False           False            -1    ********   
6         False        False           False            -1    ********   
7         False        False           False            -1    ********   
8         False        False           False            -1    ********   
9         False        False           False            -1    ********   
10        False        False           False            -1    ********   
11        False        False           False            -1    ********   
12         True         True            True            -1    ********   

   rolvaliduntil  rolbypassrls rolconfig   oid  
0           None         False      None  6171  
1           None         False      None  6181  
2           None         False      None  6182  
3           None         False      None  3373  
4           None         False      None  3374  
5           None         False      None  3375  
6           None         False      None  3377  
7           None         False      None  4569  
8           None         False      None  4570  
9           None         False      None  4571  
10          None         False      None  4200  
11          None         False      None  4544  
12          None          True      None    10

In [30]:
pg_classes_df = execute_query(
    query="""
        SELECT *
        FROM pg_class;
    """
)
pg_classes_df

oid                           relname  relnamespace  reltype  \
0    18312                     geometry_dump          2200    18314   
1    19439         idx_tiger_featnames_lname         19226        0   
2    18323                    pg_toast_18319            99        0   
3    18324              pg_toast_18319_index            99        0   
4    19440  idx_tiger_featnames_tlid_statefp         19226        0   
..     ...                               ...           ...      ...   
588  19584                 pagc_rules_id_seq         19226        0   
589  19590                    pg_toast_19585            99        0   
590  19591              pg_toast_19585_index            99        0   
591  19592                   pagc_rules_pkey         19226        0   
592  19585                        pagc_rules         19226    19587   

     reloftype  relowner  relam  relfilenode  reltablespace  relpages  \
0            0        10      0            0              0         0   
1            0        10    403        19439              0         1   
2            0        10      2        18323              0         0   
3            0        10    403        18324              0         1   
4            0        10    403        19440              0         1   
..         ...       ...    ...          ...            ...       ...   
588          0        10      0        19584              0         1   
589          0        10      2        19590              0         0   
590          0        10    403        19591              0         1   
591          0        10    403        19592              0        27   
592          0        10      2        19585              0        75   

     reltuples  relallvisible  reltoastrelid  relhasindex  relisshared  \
0         -1.0              0              0        False        False   
1          0.0              0              0        False        False   
2         -1.0              0              0         True        False   
3          0.0              0              0        False        False   
4          0.0              0              0        False        False   
..         ...            ...            ...          ...          ...   
588        1.0              0              0        False        False   
589       -1.0              0              0         True        False   
590        0.0              0              0        False        False   
591     4354.0              0              0        False        False   
592     4354.0             75          19590         True        False   

    relpersistence relkind  relnatts  relchecks  relhasrules  relhastriggers  \
0                p       c         2          0        False           False   
1                p       i         1          0        False           False   
2                p       t         3          0        False           False   
3                p       i         2          0        False           False   
4                p       i         2          0        False           False   
..             ...     ...       ...        ...          ...             ...   
588              p       S         3          0        False           False   
589              p       t         3          0        False           False   
590              p       i         2          0        False           False   
591              p       i         1          0        False           False   
592              p       r         3          0        False           False   

     relhassubclass  relrowsecurity  relforcerowsecurity  relispopulated  \
0             False           False                False            True   
1             False           False                False            True   
2             False           False                False            True   
3             False           False                False            True   
4             False           False            

In [32]:
pg_classes_df.loc[pg_classes_df["relname"].str.contains("geometry")]

oid           relname  relnamespace  reltype  reloftype  relowner  \
0   18312     geometry_dump          2200    18314          0        10   
9   18857  geometry_columns          2200    18859          0        10   
23  19083      topogeometry         19052    19085          0        10   

    relam  relfilenode  reltablespace  relpages  reltuples  relallvisible  \
0       0            0              0         0       -1.0              0   
9       0            0              0         0       -1.0              0   
23      0            0              0         0       -1.0              0   

    reltoastrelid  relhasindex  relisshared relpersistence relkind  relnatts  \
0               0        False        False              p       c         2   
9               0        False        False              p       v         7   
23              0        False        False              p       c         4   

    relchecks  relhasrules  relhastriggers  relhassubclass  relrowsecurity  \
0           0        False           False           False           False   
9           0         True           False           False           False   
23          0        False           False           False           False   

    relforcerowsecurity  relispopulated relreplident  relispartition  \
0                 False            True            n           False   
9                 False            True            n           False   
23                False            True            n           False   

    relrewrite relfrozenxid relminmxid  \
0            0            0          0   
9            0            0          0   
23           0            0          0   

                                              relacl reloptions relpartbound  
0                                               None       None         None  
9   {db_username=arwdDxt/db_username,=r/db_username}       None         None  
23                                              None       None         None

In [34]:
pg_data_types_df = execute_query(
    query="""
        SELECT *
        FROM pg_type;
    """
)
pg_data_types_df

oid      typname  typnamespace  typowner  typlen  typbyval typtype  \
0       16         bool            11        10       1      True       b   
1       17        bytea            11        10      -1     False       b   
2       18         char            11        10       1      True       b   
3       19         name            11        10      64     False       b   
4       20         int8            11        10       8      True       b   
..     ...          ...           ...       ...     ...       ...     ...   
714  19566    _pagc_gaz         19226        10      -1     False       b   
715  19577     pagc_lex         19226        10      -1     False       c   
716  19576    _pagc_lex         19226        10      -1     False       b   
717  19587   pagc_rules         19226        10      -1     False       c   
718  19586  _pagc_rules         19226        10      -1     False       b   

    typcategory  typispreferred  typisdefined typdelim  typrelid  \
0             B            True          True        ,         0   
1             U           False          True        ,         0   
2             Z           False          True        ,         0   
3             S           False          True        ,         0   
4             N           False          True        ,         0   
..          ...             ...           ...      ...       ...   
714           A           False          True        ,         0   
715           C           False          True        ,     19575   
716           A           False          True        ,         0   
717           C           False          True        ,     19585   
718           A           False          True        ,         0   

                    typsubscript  typelem  typarray   typinput   typoutput  \
0                              -        0      1000     boolin     boolout   
1                              -        0      1001    byteain    byteaout   
2                              -        0      1002     charin     charout   
3    raw_array_subscript_handler       18      1003     namein     nameout   
4                              -        0      1016     int8in     int8out   
..                           ...      ...       ...        ...         ...   
714      array_subscript_handler    19567         0   array_in   array_out   
715                            -        0     19576  record_in  record_out   
716      array_subscript_handler    19577         0   array_in   array_out   
717                            -        0     19586  record_in  record_out   
718      array_subscript_handler    19587         0   array_in   array_out   

      typreceive      typsend typmodin typmodout        typanalyze typalign  \
0       boolrecv     boolsend        -         -                 -        c   
1      bytearecv    byteasend        -         -                 -        i   
2       charrecv     charsend        -         -                 -        c   
3       namerecv     namesend        -         -                 -        c   
4       int8recv     int8send        -         -                 -        d   
..           ...          ...      ...       ...               ...      ...   
714   array_recv   array_send        -         -  array_typanalyze        d   
715  record_recv  record_send        -         -                 -        d   
716   array_recv   array_send        -         -  array_typanalyze        d   
717  record_recv  record_send        -         -                 -        d   
718   array_recv   array_send        -         -  array_typanalyze        d   

    typstorage  typnotnull  typbasetype  typtypmod  typndims  typcollation  \
0            p       False            0         -1         0             0   
1            x       False            0         -1         0             0   
2            p       False            0         -1         0             0   
3            p       False            0         -1         0       

In [36]:
pg_data_types_df.loc[pg_data_types_df["typname"].str.contains("geometry")]

oid            typname  typnamespace  typowner  typlen  typbyval  \
613  18012           geometry          2200        10      -1     False   
614  18020          _geometry          2200        10      -1     False   
623  18314      geometry_dump          2200        10      -1     False   
624  18313     _geometry_dump          2200        10      -1     False   
633  18859   geometry_columns          2200        10      -1     False   
634  18858  _geometry_columns          2200        10      -1     False   
639  19085       topogeometry         19052        10      -1     False   
640  19084      _topogeometry         19052        10      -1     False   

    typtype typcategory  typispreferred  typisdefined typdelim  typrelid  \
613       b           U           False          True        :         0   
614       b           A           False          True        :         0   
623       c           C           False          True        ,     18312   
624       b           A           False          True        ,         0   
633       c           C           False          True        ,     18857   
634       b           A           False          True        ,         0   
639       c           C           False          True        ,     19083   
640       b           A           False          True        ,         0   

                typsubscript  typelem  typarray     typinput     typoutput  \
613                        -        0     18020  geometry_in  geometry_out   
614  array_subscript_handler    18012         0     array_in     array_out   
623                        -        0     18313    record_in    record_out   
624  array_subscript_handler    18314         0     array_in     array_out   
633                        -        0     18858    record_in    record_out   
634  array_subscript_handler    18859         0     array_in     array_out   
639                        -        0     19084    record_in    record_out   
640  array_subscript_handler    19085         0     array_in     array_out   

        typreceive        typsend            typmodin            typmodout  \
613  geometry_recv  geometry_send  geometry_typmod_in  geometry_typmod_out   
614     array_recv     array_send  geometry_typmod_in  geometry_typmod_out   
623    record_recv    record_send                   -                    -   
624     array_recv     array_send                   -                    -   
633    record_recv    record_send                   -                    -   
634     array_recv     array_send                   -                    -   
639    record_recv    record_send                   -                    -   
640     array_recv     array_send                   -                    -   

           typanalyze typalign typstorage  typnotnull  typbasetype  typtypmod  \
613  geometry_analyze        d          m       False            0         -1   
614  array_typanalyze        d          x       False            0         -1   
623                 -        d          x       False            0         -1   
624  array_typanalyze        d          x       False            0         -1   
633                 -        d          x       False            0         -1   
634  array_typanalyze        d          x       False            0         -1   
639                 -        d          x       False            0         -1   
640  array_typanalyze        d          x       False            0         -1   

     typndims  typcollation typdefaultbin typdefault typacl  
613         0             0          None       None   None  
614         0             0          None       None   None  
623         0             0          None       None   None  
624         0             0          None       None   None  
633         0             0          None       None   None  
634         0             0          None       None   None  
639         0             0          None       None   None  
640         0 

[^1] Before you can use docker and compose, you have to install [docker](https://docs.docker.com/engine/install/) and [compose](https://docs.docker.com/compose/install/). I installed the docker engine and compose, but it looks like the developers at Docker Inc. guide people towards installing their Docker Desktop client. In either setup, docker should still become available to you via the command line, so instructions in this post should work.

[^2] Docker containers are designed to be a replicable instance of an image, so when you shut down your application, your containers are removed and new ones are created next time you start it up. This is great for reproducibility (you always get a new, clean instance based on your image), but you don't want the data you ingest into your database to get wiped every time you shut down your system, so you can define a persistent **volume** that will live on in the host system.